chain/chain/src/chain.rs

```diff
@@ -2891,98 +2900,187 @@ impl<'a> ChainUpdate<'a> {
-                            &receipts,
-                            chunk.transactions(),
-                            chunk_inner.validator_proposals(),
-                            prev_block.header().gas_price(),
-                            gas_limit,
-                            &block.header().challenges_result(),
-                            *block.header().random_value(),
-                            true,
-                            is_first_block_with_chunk_of_version,
-                            #[cfg(feature = "sandbox")]
-                            self.states_to_patch.take(),
-                            #[cfg(not(feature = "sandbox"))]
-                            None,
-                        )
-                        .map_err(|e| ErrorKind::Other(e.to_string()))?;
-
-                    let (outcome_root, outcome_paths) =
-                        ApplyTransactionResult::compute_outcomes_proof(&apply_result.outcomes);
-
-                    self.chain_store_update.save_trie_changes(apply_result.trie_changes);
-                    // Save state root after applying transactions.
-                    self.chain_store_update.save_chunk_extra(
-                        &block.hash(),
-                        shard_id,
-                        ChunkExtra::new(
-                            &apply_result.new_root,
-                            outcome_root,
-                            apply_result.validator_proposals,
-                            apply_result.total_gas_burnt,
-                            gas_limit,
-                            apply_result.total_balance_burnt,
-                        ),
-                    );
-                    self.chain_store_update.save_outgoing_receipt(
-                        &block.hash(),
-                        shard_id,
-                        apply_result.receipt_result,
-                    );
-                    // Save receipt and transaction results.
-                    self.chain_store_update.save_outcomes_with_proofs(
-                        &block.hash(),
-                        shard_id,
-                        apply_result.outcomes,
-                        outcome_paths,
-                    );
+                    let runtime_adapter = self.runtime_adapter.clone();
+                    let block_hash = block.hash().clone();
+                    let challenges_result = block.header().challenges_result().clone();
+                    let block_timestamp = block.header().raw_timestamp();
+                    let gas_price = prev_block.header().gas_price();
+                    let random_seed = *block.header().random_value();
+                    let height = chunk_header.height_included();
+                    let prev_block_hash = chunk_header.prev_block_hash().clone();
+                    #[cfg(feature = "sandbox")]
+                    let states_to_patch = self.states_to_patch.take();
+
+                    same_height_handlers.push(thread::spawn(
```

Comment:
I don't think spawning the threads directly is necessary the best approach. For example, Meyers effective modern C++ advises against that: Item 35: Prefer task-based programming to thread-based.
 
A specific problems we can face are

* threads are somewhat costly to create, for best performance, we want to cache and re-use threads
* this spawns the number of threads proportional to the amount of work, and the work is CPU bound. That means that we can easily over-subscribe the CPU and spend unnecessary time on context switches. It's better to spawn the number of threads proportional to the number of CPUs
* threads consume limited OS resources. If you spawn more than 10k threads, you can run into a situation where `thread::spawn` just panics. Note that some of those resources are global per OS. I don't think will run into the those limits (they are pretty high in practice today), but it feels more robust to not have potentially unbounded resource usage.

If we pursue this approach with imperative task spawning, I suggest using a thread pool to control the OS threads we use.

Reply:
Ok, you've persuaded me, it would be no harder to switch from rayon to something else in the future if need arises

Reply:
To clarify, I am not married to rayon in particular: hand-rolled threadpool would work fine as well. In rust-analyzer, I use https://docs.rs/threadpool which is a dead simple thing. 

One problem with rayon is that its a pretty complex beast actually. https://github.com/sorbet/sorbet/blob/ba614a5a69c77c89c88b8c9983185927856b6ac0/common/concurrency/WorkerPool.h shows that this thing needn't be complex.